### PR TITLE
Add TimSort Algorithm

### LINKED ISSUE #16: Add a `.cpp` implementation for `Tim Sort`.
- [ ] Add a `.cpp` implementation for `Tim Sort`.
- [ ] Create folder for the same under `sorting`.

Idea from https://github.com/Ashrockzzz2003/Data_Structures_and_Algorithms/issues/8. Reserved for @adofm

### PR DESCRIPTION
**Pull request by adofm**
**Description:**

This PR introduces the TimSort algorithm, a hybrid sorting method combining insertion sort and merge sort for efficient and stable sorting. The implementation divides the array into small runs, sorts them using insertion sort, and merges them iteratively.

**Screenshot**

image

Tested and works as expected.
Fixes #16

### CODE REVIEW

File: sorting/tim_sort/tim_sort.cpp [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+// insertion sort
+void perfInsSort(int arr[], int start, int stop)
+{
+    for (int idx = start + 1; idx <= stop; ++idx)
+    {
+        int current = arr[idx];
+        int prev = idx - 1;
+        while (prev >= 0 && arr[prev] > current)
```
Comment (FirefoxSRV) [inline sorting/tim_sort/tim_sort.cpp:38]: This logic seems to be wrong.....try for all edge cases ( array sizes more than 32 or 64 ) and comment an output screenshot below this pr comment

**Review by FirefoxSRV — CHANGES_REQUESTED** [review]
Leave a screenshot of array sizes greater than 32 or 64.

File: sorting/tim_sort/tim_sort.cpp (hunk 1 above)
Comment (adofm) [reply sorting/tim_sort/tim_sort.cpp:38]: my bad it should be prev>=start
will change it and commit

**Comment by adofm** [comment]
Array Size 128 


image


Array Size 31


image


Array Size 0 



image

**Review by FirefoxSRV — APPROVED** [review]

File: sorting/tim_sort/tim_sort.cpp (hunk 1 above)
Comment (FirefoxSRV) [reply sorting/tim_sort/tim_sort.cpp]: Yes, good eyes....keep it up

**Review by FirefoxSRV — APPROVED** [review]
Sorting now works for all cases

**Comment by Ashrockzzz2003** [comment]
Considering other PRs
Bonus 20 points for @adofm if you can format your code comments.

1. Leave one char space after comment token `//`
2. Comments are complete English sentences well punctuated. Upper case and lower case must be proper.
3. End comments with a full stop.
4. Self-Explanatory variable names

**Comment by adofm** [comment]
Hi ,
@Ashrockzzz2003 I will take this up and give proper documentation as you mentioned.

**Comment by adofm** [comment]
@Ashrockzzz2003 
Should I create a new PR for this, or should I continue with the existing one?

**Pull request merged**